solver/jobs.go

```diff
@@ -632,6 +632,7 @@ func (s *sharedOp) CacheMap(ctx context.Context, index int) (resp *cacheMapResp,
 		}
 		ctx = opentracing.ContextWithSpan(progress.WithProgress(ctx, s.st.mpw), s.st.mspan)
 		ctx = session.NewContext(ctx, s.st.getSessionID())
+		ctx = withAncestorCacheOpts(ctx, s.st)
```

Comment:
Okay, I'll take a closer look at #1551 and give integrating with the Group a shot

Reply:
Alright, took a look, my understanding is the issue comes down to the fact that we construct a fetcher [here](https://github.com/sipsma/buildkit/blob/32e67c439cf08881a91053ce00d62be6bdf7c88c/util/pull/pull.go#L89) from the resolver, at which time `credentials` [will be called with whatever groups are present at the time](https://github.com/sipsma/buildkit/blob/94e8ab5a3b104d6ba7884375e2d8f391ceb3b1e5/util/resolver/resolver.go#L170). That's bad because it could be much later and under a different vtx that the actual de-lazy happens but the provider [will still use that possibly stale fetcher](https://github.com/sipsma/buildkit/blob/94e8ab5a3b104d6ba7884375e2d8f391ceb3b1e5/util/pull/pull.go#L200).

If that's correct then I guess the fix is, in part, to recreate the fetcher from the resolver when the actual pull occurs rather than use the one created earlier, which is easy enough. My only question is if you think it's okay at that time (i.e. [this line](https://github.com/sipsma/buildkit/blob/94e8ab5a3b104d6ba7884375e2d8f391ceb3b1e5/util/pull/pull.go#L200)) to use the same resolver as before (which should be updated w/ sessions in calls to `CacheMap` and `Snapshot` such as [here](https://github.com/sipsma/buildkit/blob/94e8ab5a3b104d6ba7884375e2d8f391ceb3b1e5/source/containerimage/pull.go#L253-L257)), or do you think we need to make a new resolver using the `session.Group` from the vtx where the actual de-lazy is occurring (i.e. [here](https://github.com/sipsma/buildkit/blob/94e8ab5a3b104d6ba7884375e2d8f391ceb3b1e5/cache/refs.go#L549) or in the exporter)?

Basically, should the de-lazy use the `session.Group`s from the original vtx where the lazy ref got created, or the `session.Group` from the vtx where the de-lazy is occurring? Or is it irrelevant (I haven't traced through #1551 carefully enough yet to tell)?

Reply:
> My only question is if you think it's okay at that time (i.e. this line) to use the same resolver as before

Yes, ideally we should reuse as much as possible. New resolver means new tcp connections and new authentication roundtrips that should be avoided.

> Basically, should the de-lazy use the session.Groups from the original vtx where the lazy ref got created, or the session.Group from the vtx where the de-lazy is occurring?

From where the de-lazy is occurring. Old may not be valid anymore.

I added this `SessionAuthenticator` https://github.com/moby/buildkit/blob/master/source/containerimage/pull.go#L127 concept that allows adding new sessions while reusing the resolver https://github.com/moby/buildkit/blob/master/source/containerimage/pull.go#L167 . Was actually thinking to take this further and write a custom containerd authenticator and resolver pool (limited resolver pool already exists for pull) that can remember tokens and reuse connections cross builds(after checking auth). Something like `resolverpool.Get(ref, scope ("pull/push"), g session.Group)`. (Maybe also `resolver.WithSession(session.Group) resolver` if that helps). That could help simplify this as well. Maybe I have time to work on this tomorrow.

Reply:
Okay, for now all I did was make it so the fetcher gets [reconstructed at the time of de-lazy](https://github.com/sipsma/buildkit/blob/c38c3cf32d2c7d018a94078372312af825f24473/util/pull/pull.go#L192-L195).

But it is still using the resolver that is only ever [updated w/ `session.Group`s from calls to `CacheKey`/`Snapshot`](https://github.com/sipsma/buildkit/blob/c38c3cf32d2c7d018a94078372312af825f24473/source/containerimage/pull.go#L176-L180). So it will not be updated  with the `session.Group`s from the vertex where actual de-lazy is occurring. I'm not sure what the best way to implement that would be as it seems like it would require plumbing the `session.Group`s for the current vtx all the way through to the provider implementation, where it can be used to update the resolver. I guess we could put the session group in the `ctx`, but it seems like that was something you were trying to avoid in the previous PR? Let me know what you think or if I'm misunderstanding.